File: back_end/node_server/src/routes/job_detail.js
// src/routes/job_detail.js
const express = require("express");
const { getDb, toObjectId } = require("../db/mongo");
const { requireAuth } = require("../middleware/auth");
const { isAdmin } = require("../middleware/authz");

const router = express.Router();

const JOBS_COLLECTION = process.env.JOBS_COLLECTION || "jobs";
const PROGRESS_COLLECTION = process.env.PROGRESS_COLLECTION || "progress";

function mapDoc(doc) {
  if (!doc) return null;
  const { _id, ...rest } = doc;
  return { id: _id.toString(), ...rest };
}

/**
 * @swagger
 * /api/jobs/{id}/full:
 *   get:
 *     summary: Job 상세 + Progress 통합 조회
 *     description: 특정 job의 상세 정보와 progress 정보를 함께 조회합니다.
 *     tags: [Job Detail]
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         schema:
 *           type: string
 *     responses:
 *       200:
 *         description: Job 상세 + Progress 조회 성공
 */
router.get("/jobs/:id/full", requireAuth, async (req, res) => {
  const { id } = req.params;

  let objectId;
  try {
    objectId = toObjectId(id);
  } catch {
    return res.status(400).json({ error: "invalid job id" });
  }

  try {
    const db = getDb();
    const jobsCol = db.collection(JOBS_COLLECTION);
    const progressCol = db.collection(PROGRESS_COLLECTION);

    const jobDoc = await jobsCol.findOne({ _id: objectId });
    if (!jobDoc) {
      return res.status(404).json({ error: "job not found" });
    }

    if (!isAdmin(req)) {
      const owner = jobDoc?.ownerUserId || jobDoc?.owner_id;
      if (!owner || String(owner) !== String(req.user.id)) {
        return res.status(403).json({ error: "Forbidden" });
      }
    }

    // FastAPI에서 progress.job_id 를 문자열 job_id 로 저장했다고 가정
    const progressDocs = await progressCol
      .find({ job_id: id })
      .sort({ created_at: 1 })
      .toArray();

    res.json({
      job: mapDoc(jobDoc),
      progress: progressDocs.map(mapDoc),
    });
  } catch (err) {
    console.error(
      "GET /api/jobs/:id/full error:",
      err.message
    );
    res.status(500).json({ error: "failed to get full job info" });
  }
});

module.exports = router;

File: back_end/fast_server/endpoints/job_detail.py
from fastapi import APIRouter, Depends, HTTPException

from bson import ObjectId

from db import jobs_col, progress_col
from endpoints.auth import get_current_user, is_admin

router = APIRouter(tags=["job_detail"])


def _oid(job_id: str) -> ObjectId:
    try:
        return ObjectId(job_id)
    except Exception:
        raise HTTPException(status_code=400, detail="invalid job id")


@router.get("/jobs/{job_id}")
def get_job(job_id: str, user=Depends(get_current_user)):
    doc = jobs_col.find_one({"_id": _oid(job_id)})
    if not doc:
        raise HTTPException(status_code=404, detail="not found")
    if not is_admin(user):
        owner = doc.get("ownerUserId") or doc.get("owner_id") or doc.get("ownerId")
        if not owner or str(owner) != str(user["id"]):
            raise HTTPException(status_code=403, detail="forbidden")
    doc["id"] = str(doc.pop("_id"))
    return doc


@router.get("/jobs/{job_id}/progress")
def get_job_progress(job_id: str, user=Depends(get_current_user)):
    # Prefer the canonical progress doc (_id == job_id) to avoid mixed-type sorting issues.
    progress = progress_col.find_one({"_id": _oid(job_id)})
    if not progress:
        progress = progress_col.find_one({"job_id": job_id}, sort=[("updated_at", -1)])
    if not progress:
        return {
            "job_id": job_id,
            "status": "queued",
            "progress": 0.0,
            "epoch": 0,
            "total_epochs": 0,
            "logs": [],
        }
    if not is_admin(user):
        owner = progress.get("ownerUserId") or progress.get("owner_id") or progress.get("ownerId")
        if owner and str(owner) != str(user["id"]):
            raise HTTPException(status_code=403, detail="forbidden")
        if not owner:
            job = jobs_col.find_one({"_id": _oid(job_id)})
            if job:
                job_owner = job.get("ownerUserId") or job.get("owner_id") or job.get("ownerId")
                if job_owner and str(job_owner) != str(user["id"]):
                    raise HTTPException(status_code=403, detail="forbidden")
            else:
                # job이 없어도 최소한의 진행 정보는 빈 값으로 내려보낸다
                return {
                    "job_id": job_id,
                    "status": "queued",
                    "progress": 0.0,
                    "epoch": 0,
                    "total_epochs": 0,
                    "logs": [],
                }
    return progress
